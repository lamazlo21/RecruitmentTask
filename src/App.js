import React, { Component } from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';

// Components
import Form from './components/Form';

// Styles
import '../public/styles/styles.css';
import axios from "axios";

class App extends Component {

  state={
    errors: []
  };


    validEmail=(email)=>{
        return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
    }

    validDate=(date)=>{
        let dateArr = [];
        let currDate = new Date();
        let currYear = currDate.getFullYear();
        let currMonth = currDate.getMonth()+1;
        let currDay = currDate.getDate();
        dateArr = date.split('-');
        for(let i in dateArr){
            dateArr[i] = parseInt(dateArr[i],10)
        }
        if(dateArr[0]>=currYear&&dateArr[1]>=currMonth&&dateArr[2]>=currDay)
            return true
        return false
    }

    postMethod=(url, name, surname, email, date, array)=>{
        axios.post(url,{
            name: name.value,
            surname: surname.value,
            email: email.value,
            date: date.value
        }).then((res) => {
            array.push(res.data)
            this.setState({errors: array},()=>{
                name.value ='';
                surname.value ='';
                email.value='';
                date.value = '';
            })
        }).catch(res => {
            if(res.data===undefined)
                array.push({
                    data: "There is a problem connecting to the server!",
                    type: false
                })
            else
                array.push(res.data)
            this.setState({errors: array})
        })

    }

    sendForm=(e)=> {
        e.preventDefault();
        let array = [];
        let name = document.getElementById('inputName');
        let surname = document.getElementById('inputSurname');
        let email = document.getElementById('inputEmail');
        let date = document.getElementById('inputDate');
        //Basic validation
        if(name.value.length<1||name.value.length>50) {
            array.push({
                data: "Name has to contain between 1 to 50 letters!",
                type: false
            })
        }
        if(surname.value.length<1||surname.value.length>50) {
            array.push({
                data: "Surname has to contain between 1 to 50 letters!",
                type: false
            })
        }
        if(!this.validEmail(email.value)){
            array.push({
                data:"Email format is incorrect!",
                type:false
            })
        }
        if(!(date.value.length>0&&this.validDate(date.value))){
            array.push({
                data:"You cannot choose a date from the past!",
                type:false
            })
        }

        if(array.length===0)
            this.postMethod('http://127.0.0.1:4000/', name, surname, email, date, array);
        this.setState({errors: array})
    }

     cleanAlerts=()=>{
        this.setState({errors: []});
    }


    render() {
    let errors = this.state.errors.map((val, index)=>{
      return(
          //The alert is green when form is accepted and red when the error occurs
      <li className={val.type===true?"alert alert-success box":"alert alert-danger box"}  role="alert" key={index}>
          {val.data}
      </li>
      );
    })
    return (
      <div className="App">
        {errors}
        <Form sendForm={this.sendForm} cleanAlerts={this.cleanAlerts}/>
      </div>
    );
  }
}



export default (App);