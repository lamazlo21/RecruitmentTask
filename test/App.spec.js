import React from 'react';
import App from '../src/App';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import mockAxios from 'axios';


describe('Component test', ()=> {
    it('Renders App without crash', () => {
        shallow(<App/>);
    })
    it('Renders list of errors based on state.errors',()=>{
        const app = shallow(<App/>);
        app.setState({
            errors: [
                { data:"Error", type:false},
                {data:"Correct",type:true}
            ]
        });
        expect(toJson(app)).toMatchSnapshot();
    })
})



describe('Form validation', () => {
    describe('Email Validation',()=>{
        it('Should return true',()=>{
            const wrapper = shallow(<App/>)
            const instance = wrapper.instance();
            let email1 = instance.validEmail('lamazlo21@wp.pl');
            let email2 = instance.validEmail('lam.zlo@wp.pl');
            expect(email1).toBeTruthy();
            expect(email2).toBeTruthy();
        })
        it('Should return false',()=>{
            const wrapper = shallow(<App/>)
            const instance = wrapper.instance();
            let email1 = instance.validEmail('lam@zlo@wp.pl');
            let email2 = instance.validEmail('lam@zlo@wp.');
            let email3 = instance.validEmail('lam@zlo@wppl');
            let email4 = instance.validEmail('lamwp.pl');
            let email5 = instance.validEmail('@');
            expect(email1).toBeFalsy();
            expect(email2).toBeFalsy();
            expect(email3).toBeFalsy();
            expect(email4).toBeFalsy();
            expect(email5).toBeFalsy();
        })
    })

    describe('Date Validation',()=>{
        it('Should return true',()=>{
            const wrapper = shallow(<App/>)
            const instance = wrapper.instance();
            let date1 = instance.validDate('10-16-2019');
            let date2 = instance.validDate('11-10-2019');
            let date3 = instance.validDate('05-05-2020');
            expect(date1).toBeTruthy;
            expect(date2).toBeTruthy;
            expect(date3).toBeTruthy;
        })
        it('Should return false',()=>{
            const wrapper = shallow(<App/>)
            const instance = wrapper.instance();
            let date1 = instance.validDate('10-16-2018');
            let date2 = instance.validDate('10-10-2019');
            let date3 = instance.validDate('05-05-2019');
            expect(date1).toBeFalsy;
            expect(date2).toBeFalsy;
            expect(date3).toBeFalsy;
        })

    })
});

describe('Methods validation',()=>{
    describe('cleanAlerts method validation',()=>{
        it('cleanAlerts cleans errors.state', ()=>{
            const app = shallow(<App/>);
            const instance = app.instance();
            app.setState({
                errors: [{
                    data: "error",
                    type: false
                }]
            })
            expect(app.state('errors')).toHaveLength(1);
            instance.cleanAlerts();
            expect(app.state('errors')).toHaveLength(0);
        })
    })

})



