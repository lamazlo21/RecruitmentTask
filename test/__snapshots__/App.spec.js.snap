exports[`Component test Renders list of errors based on state.errors 1`] = `
<div
  className="App">
  <li
    className="alert alert-danger box"
    key="0"
    role="alert">
    Error
  </li>
  <li
    className="alert alert-success box"
    key="1"
    role="alert">
    Correct
  </li>
  <Form
    cleanAlerts={[Function]}
    sendForm={[Function]} />
</div>
`;
